### ISSUE TITLE
关于Linear Projection

### ISSUE THREAD

**Issue by aslicedbread** [ProHiryu/transfer-learning-for-chinese-ner#1]
您好，您是否可以解释一下Linear Projection的原理，我参考Multi-task Domain Adaptation for Sequence Tagging中的说明，不是很理解。另，Transfer Learning for Sequence Tagging with Hierarchical Recurrent Networks中应该是没有提Domain Projections？

**Comment by ProHiryu** [comment]
对，linear的意思就是直接使用线性网络对输入到crf网络参数进行处理，我这边直接使用crf_log_likelihood函数取得crf结果，而domain mask在论文Multi-task Domain Adaptation for Sequence Tagging有比较详细的解释，可以理解为用mask将不同domain的数据分隔开，让模型一次只关注一个domain的数据

**Comment by aslicedbread** [comment]
> 对，linear的意思就是直接使用线性网络对输入到crf网络参数进行处理，我这边直接使用crf_log_likelihood函数取得crf结果，而domain mask在论文Multi-task Domain Adaptation for Sequence Tagging有比较详细的解释，可以理解为用mask将不同domain的数据分隔开，让模型一次只关注一个domain的数据

谢谢您的解答，我还是有一些疑问，关于代码中的project_layer的作用，在project_layer中，对BiLSTM的输出作了如下操作，hidden = tf.tanh(tf.nn.xw_plus_b(lstm_output, w, b))，logits = tf.nn.xw_plus_b(hidden, w, b)，目的是为了什么呢？
此外，关于Domain Projections中的Domain Masks，论文中作者解释The mask m_d has value 1 for the effective dimensions of domain d and domain shared region, and 0 for all other dimensions，h^ = m_d ⊙ h，其中We can then apply these masks directly to the hidden vectors h learned by the BiLSTM to produce a projected hidden state h^，那这里h是学习了一个domain的数据集得到的吗？有点不理解，如果是的话，那么h应该是不能学习到两个domain（假设只有2个domain）shared部分，或者说m_d是怎么获得的呢，它是如何知道哪一部分是两个domain shared，哪一部分是各自独享的？因此在代码中，start_logits、end_logits、pad_logits分别是用来表示什么含义，我不是很理解，还希望您能解答一下，谢谢。

**Comment by ProHiryu** [comment]
首先第一个问题：

对于linear proj来说，原论文已经说了使用一个k维的矩阵来直接加到我们产生的output上，这里直接做了一个单层hidden的感知器，用于获取输出，一般来说 nlp应用也可以直接获取lstm的output直接设置一层output layer得到输出，这里用了两层

然后是论文中说的h是指lstm层输出的output，是这样的，关于shared这一层，作者是假设前三分之一的参数是学习二者共同的，中间三分之一是domain1，最后三分之一是Domain2，对于神经网络来说，可解释性都是人为加上的，我们就假定参数的学习是这样分布的，所以加入md来进行mask，代码中的Domain mask由于使用的是logits而不是output，所以维度会不一样，start end都是用作logits的padding，最后得到的维度是Batch_size x [time_Step + 2] x [num_tags + 2] 其中tag维度最后两个元素就表明了这是和哪个Domain有关系，其中start可以理解为只和第一个Domain有关系因为最后两个元素为[1, 0]，end相反，如果还是不理解，建议手动把所有变量的维度都写出来，应该就懂了

**Comment by aslicedbread** [comment]
> 首先第一个问题：
> 
> 对于linear proj来说，原论文已经说了使用一个k维的矩阵来直接加到我们产生的output上，这里直接做了一个单层hidden的感知器，用于获取输出，一般来说 nlp应用也可以直接获取lstm的output直接设置一层output layer得到输出，这里用了两层
> 
> 然后是论文中说的h是指lstm层输出的output，是这样的，关于shared这一层，作者是假设前三分之一的参数是学习二者共同的，中间三分之一是domain1，最后三分之一是Domain2，对于神经网络来说，可解释性都是人为加上的，我们就假定参数的学习是这样分布的，所以加入md来进行mask，代码中的Domain mask由于使用的是logits而不是output，所以维度会不一样，start end都是用作logits的padding，最后得到的维度是Batch_size x [time_Step + 2] x [num_tags + 2] 其中tag维度最后两个元素就表明了这是和哪个Domain有关系，其中start可以理解为只和第一个Domain有关系因为最后两个元素为[1, 0]，end相反，如果还是不理解，建议手动把所有变量的维度都写出来，应该就懂了


嗯我写了一遍start_logits, end_logits, pad_logits, targets的维度，代码中batch_size=20,num_targets=13,依旧有困惑：
（1）small = -1000是什么作用，或者说为何要取-1000？

以其中一个batch为例：
Project_Layer层的输出logits：
logits_test:shape+ (20, 35, 13)

start_logits_test:shape+ (20, 1, 15) [13:14] = [-1000,0]
end_logits_test:shape+ (20, 1, 15)   [13:14] = [0,-1000] # start与end在axis=-1维度最后两位不同

pad_logits_test:shape+ (20, 35, 2)

final_logits_test:shape+ (20, 37, 15)

_targets_test:shape+ (20, 35)

# 左边插入14，右边插入13，得到最终的targets
targets_test:shape+ (20, 37)

写完，我还是不理解在即便只有两个Domain情况下，Domain Mask的实现机理：

您之前回复我：代码中的Domain mask由于使用的是logits而不是output，所以维度会不一样，start end都是用作logits的padding，最后得到

的维度是Batch_size x [time_Step + 2] x [num_tags + 2]，其中tag维度最后两个元素就表明了这是和哪个Domain有关系，其中start可以理

解为只和第一个Domain有关系因为最后两个元素为[1, 0]，end相反。

（2）我不太理解，为何tag维度最后两个元素就表明了这是和哪个Domian有关系，我大致可以理解为是一种one-hot, 10表示第一个Domain，01

表示第二个Domain？此外，为何concat是[start_logits,logits,end_logits]而不是[logits,start_logits,end_logits]或者其他？另，这种Domain Mask机制在代码中是如何进行体现的，也即m_d（您之前说神经网络的可解释性是人为添加的，也包括h^ = m_d ⊙ h）？【我的理解中这种Domain Mask机制，模型通过Joint Learning学习参数，】

您能否麻烦再解释一下，谢谢。

**Comment by aslicedbread** [comment]
> 首先第一个问题：
> 
> 对于linear proj来说，原论文已经说了使用一个k维的矩阵来直接加到我们产生的output上，这里直接做了一个单层hidden的感知器，用于获取输出，一般来说 nlp应用也可以直接获取lstm的output直接设置一层output layer得到输出，这里用了两层
> 
> 然后是论文中说的h是指lstm层输出的output，是这样的，关于shared这一层，作者是假设前三分之一的参数是学习二者共同的，中间三分之一是domain1，最后三分之一是Domain2，对于神经网络来说，可解释性都是人为加上的，我们就假定参数的学习是这样分布的，所以加入md来进行mask，代码中的Domain mask由于使用的是logits而不是output，所以维度会不一样，start end都是用作logits的padding，最后得到的维度是Batch_size x [time_Step + 2] x [num_tags + 2] 其中tag维度最后两个元素就表明了这是和哪个Domain有关系，其中start可以理解为只和第一个Domain有关系因为最后两个元素为[1, 0]，end相反，如果还是不理解，建议手动把所有变量的维度都写出来，应该就懂了

嗯我写了一遍start_logits, end_logits, pad_logits, targets的维度，代码中batch_size=20,num_targets=13,依旧有困惑：
（1）small = -1000是什么作用，或者说为何要取-1000？

以其中一个batch为例：
Project_Layer层的输出logits：
logits_test:shape+ (20, 35, 13)

start_logits_test:shape+ (20, 1, 15) [13:14] = [-1000,0]
end_logits_test:shape+ (20, 1, 15)   [13:14] = [0,-1000]  # start与end在axis=-1维度最后两位不同

pad_logits_test:shape+ (20, 35, 2)

final_logits_test:shape+ (20, 37, 15)

_targets_test:shape+ (20, 35)

左边插入14，右边插入13，得到最终的targets
targets_test:shape+ (20, 37)

写完，我还是不理解在即便只有两个Domain情况下，Domain Mask的实现机理：

您之前回复我：代码中的Domain mask由于使用的是logits而不是output，所以维度会不一样，start end都是用作logits的padding，最后得到

的维度是Batch_size x [time_Step + 2] x [num_tags + 2]，其中tag维度最后两个元素就表明了这是和哪个Domain有关系，其中start可以理

解为只和第一个Domain有关系因为最后两个元素为[1, 0]，end相反。

（2）我不太理解，为何tag维度最后两个元素就表明了这是和哪个Domian有关系，我大致可以理解为是一种one-hot, 10表示第一个Domain，01表示第二个Domain？此外，为何concat是[start_logits,logits,end_logits]而不是[logits,start_logits,end_logits]或者其他？另，这种Domain Mask机制在代码中是如何进行体现的，也即m_d（您之前说神经网络的可解释性是人为添加的，也包括h^ = m_d ⊙ h）？

您能否麻烦再解释一下，谢谢。

**Comment by ProHiryu** [comment]
1）使用-1000是为了更好的区分二者，也可以使用0，我个人认为使用0应该也可以达到不错的效果

2）当对输出进行了mask之后，神经网络就会自动对二者进行区分，输入到index=tag+1与index=tag+2的参数就会不同，进而在做推理的时候影响到结果，相当于加了两个分类tag来区别迁移学习的二者，至于concat，只要input和labels保持前后一致，当然也可以使用你说的那个顺序
